### ISSUE TITLE
Soft spans

### ISSUE THREAD

**Issue by rkusa** [google/xi-editor#244]
Since the PR #230 is not ready yet, I'd like to move the discussion here.

For some context, here an actual use case of soft spans.

> ![2017-04-25 16 58 24](https://cloud.githubusercontent.com/assets/409021/25394596/c57cd650-29de-11e7-8f30-9a8fd09210bc.gif)

> In this example the arrow keys are not used to move the cursor over the closing parenthesis, a closing parenthesis is inserted instead.

In the initial PR I've implemented the following characteristics, that are up for discussion.

> ![2017-04-25 16 59 18](https://cloud.githubusercontent.com/assets/409021/25394803/55b8d818-29df-11e7-88ac-2603ffbd269c.gif)

> - if a character is inserted before a soft span and the character matches with the first character of the soft span, the cursor is shifted over this character (implementation-wise the existing value of the soft span is overridden)
> - if more than one character is inserted (e.g. using paste), it works according the point above, but shifts the cursor for as long as the input string matches the soft span string
> - if the cursor is moved (arrow keys, mouse click, ...), all soft spans before the cursor are removed
> - if text is deleted, soft spans are shifted accordingly

**When should a soft span become real?**

In my initial implementation, when moving the cursor, all soft span (parts) before the cursor became real.

@trishume added and suggested

> I imagine this is supposed to work like the equivalent feature in XCode, so I tested XCode's behaviour and found that any movement off of the line the soft close paren is on, the close paren becomes real (and XCode has a cool flash animation that calls attention to this).

> I think this behaviour makes more sense than only removing soft spans before the cursor. Even if you move up a line, I think the spans on the line you are at should become real, at least for paren matching.

At first, I also made all soft spans real, once the cursor was moved manually (in whatever direction) and tbh, I don't remember the use case for which I changed it to only make soft spans before the cursor real.

**Should it be possible to easily remove soft spans?**

Sometimes, I catch myself hitting Esc trying to remove soft spans (in other editors). Since this does not work, the next step I am subconsciously do, is to hit Undo. However, this always removes not only the soft span, but also the previously added text.
So I was wondering if Esc should be a shortcut for removing soft spans or if we should just create a new Undo group for soft spans?

---

Other comments.

@trishume also added

> Another interesting thing I noticed is that XCode only creates the soft close paren after you type a character following the opening paren. I'm not sure the reason behind this, but one thing that often annoys me about bracket matching in other text editors is when I do some editing operation that messes up paren nesting, so I go to insert a missing open paren, but then the editor adds a close paren for it! The XCode behaviour avoids doing this, which is cool.

Interesting behaviour! I would see this being a good practice of how a auto-pair plugin should behave.

> I also tested Sublime's behaviour and it seems that typing a ) with your cursor before any ) always just moves past it. Typing an opening paren in the middle of other characters doesn't insert a close paren, but typing one before whitespace in the middle of a line does.

I am also constantly annoyed by Sublime's behaviour and I think this is due to Sublime not seeing closing parens as "soft spans" and just always moves the cursor when adding a closing paran before an existing closing paren.

**Comment by cmyr** [comment]
Just a thought: 

I think the defining behaviour of soft spans is that if you type the soft character, it inserts 'overtop' of the soft character and moves the cursor.

Maybe all other behaviours (around what happens when you click, etc) should be the responsibility of the plugin? I.e. a plugin could receive the mouse click event and then remove the spans (making the text 'normal'), or it could remove the text, or anything else.

For instance: a soft closing paren should behave differently from a soft "completion suggestion"; in the first case, inserted text should be inserted to the left of the peren, and it should be repositioned; in the second case (assuming the insert != the soft span text) the text should be deleted or replaced with a new suggestion.

I am often frustrated by this behaviour, and it might be nice if plugin authors could tune the experience, and users could find what felt right to them.

**Comment by rkusa** [comment]
> Maybe all other behaviours (around what happens when you click, etc) should be the responsibility of the plugin? 

Good point that should be talked about. My first thought was also to keep the responsibility of removing soft spans at the plugin's side. However, as a plugin author I would prefer them to be fire and forget (and not have to implement the logic necessary to keep track of existing soft spans).

**Comment by cmyr** [comment]
yes, I think there could be API on the client side to make this easier. I imagine the API for instance, without any other specification there could be a default span behaviour, similar to that in your demo; and then there could be some optional flags like `CANCEL_ON_OVERWRITE` and `INSERT_ON_OVERWRITE` that could be passed to choose from a predefined behaviour; and if you wanted to do anything more complicated then that you could still access and modify the spans directly.
